src/table_builder_test.cc

```diff
@@ -400,13 +400,98 @@ TEST_F(TableBuilderTest, DictCompress) {
   std::unique_ptr<TableReader> base_reader;
   NewTableReader(base_name_, &base_reader);
   std::unique_ptr<BlobFileReader> blob_reader;
-  std::unique_ptr<RandomAccessFileReader> file;
-  NewFileReader(blob_name_, &file);
-  uint64_t file_size = 0;
-  ASSERT_OK(env_->GetFileSize(blob_name_, &file_size));
-  Status stat = BlobFileReader::Open(cf_options_, std::move(file), file_size,
-                                     &blob_reader, nullptr);
-  assert(stat.code() == Status::kNotSupported);
+  NewBlobFileReader(&blob_reader);
+
+  ReadOptions ro;
+  std::unique_ptr<InternalIterator> iter;
+  iter.reset(base_reader->NewIterator(ro, nullptr /*prefix_extractor*/,
+                                      nullptr /*arena*/, false /*skip_filters*/,
+                                      TableReaderCaller::kUncategorized));
+  iter->SeekToFirst();
+  for (char i = 0; i < n; i++) {
+    ASSERT_TRUE(iter->Valid());
+    std::string key(1, i);
+    ParsedInternalKey ikey;
+    ASSERT_TRUE(ParseInternalKey(iter->key(), &ikey));
+    ASSERT_EQ(ikey.user_key, key);
+    ASSERT_EQ(ikey.type, kTypeBlobIndex);
+    BlobIndex index;
+    ASSERT_OK(DecodeInto(iter->value(), &index));
+    ASSERT_EQ(index.file_number, kTestFileNumber);
+    BlobRecord record;
+    PinnableSlice buffer;
+    ASSERT_OK(blob_reader->Get(ro, index.blob_handle, &record, &buffer));
+    ASSERT_EQ(record.key, key);
+    ASSERT_EQ(record.value, std::string(kMinBlobSize, i));
+    iter->Next();
+  }
+  ASSERT_TRUE(!iter->Valid());
+#endif
+}
+
+TEST_F(TableBuilderTest, VanillaDecodeDictCompress) {
+#if ZSTD_VERSION_NUMBER >= 10103
+  CompressionOptions compression_opts;
+  compression_opts.enabled = true;
+  compression_opts.max_dict_bytes = 4000;
+  cf_options_.blob_file_compression_options = compression_opts;
+  cf_options_.blob_file_compression = kZSTD;
+
+  table_factory_.reset(new TitanTableFactory(
+      db_options_, cf_options_, db_impl_.get(), blob_manager_, &mutex_,
+      blob_file_set_.get(), nullptr));
+
+  std::unique_ptr<WritableFileWriter> base_file;
+  NewBaseFileWriter(&base_file);
+  std::unique_ptr<TableBuilder> table_builder;
+  NewTableBuilder(base_file.get(), &table_builder);
+
+  // Build a base table and a blob file.
+  const int n = 100;
+  for (char i = 0; i < n; i++) {
+    std::string key(1, i);
+    InternalKey ikey(key, 1, kTypeValue);
+    std::string value;
+    value = std::string(kMinBlobSize, i);
+    table_builder->Add(ikey.Encode(), value);
+  }
+  ASSERT_OK(table_builder->Finish());
+  ASSERT_OK(base_file->Sync(true));
+  ASSERT_OK(base_file->Close());
+  BlobFileExists(true);
+
+  // reset cf_options_, decode using vanilla reader should also works
```

Comment:
Why is a plain decoder able to read compressed information? Am I missing something?

Reply:
BlobFileReader will respect the header of blob files (which indicate whether the file has uncompression dictionary or not).
And the decoder is on "record" level, which means, every time we want to decode a record, we will create a decoder using the uncompression dictionary owned by the reader, it firstly decode the record header(which will tell what compression method the record use), then decode the content.so actually in the whole decoding process, the options passed in has not be used, the file reader will detect from the blob file itself.

Reply:
I see that compression type is hardcoded in blob file, so cf_options is not used by blob reader. Do you think we need to error when those two aren't consistent? @yiwu-arbug

Reply:
I see. It is okay to have inconsistency in the compression options in cf_options and blob file. The cf_options only tells what compression to use for *new* blob files.